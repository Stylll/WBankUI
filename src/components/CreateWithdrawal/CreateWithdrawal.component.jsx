import React, { useState, useEffect, useRef } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import * as accountActions from '../../redux/actionCreators/accountActions';
import TextInput from '../TextInput/TextInput.component';
import Button from '../Button/Button.component';
import Dropdown from '../Dropdown/Dropdown.component';
import { validateCreateWithdrawal } from '../../helpers/validators/accountValidator';
import { currencies } from '../../helpers/defaults';
import './CreateWithdrawal.styles.scss';

const CreateWithdrawal = ({
  createWithdrawal, isLoading: reduxIsLoading, closeModal,
  errorMessage, accountWithdrawalSuccess,
}) => {
  const [values, setValues] = useState({
    accountNo: '',
    amount: '',
    currency: '',
  });

  const [formErrors, setFormErrors] = useState({
    accountNo: '',
    amount: '',
    currency: '',
  });

  const [isLoading, setIsLoading] = useState(false);

  const [serverError, setServerError] = useState('');

  const formSubmitted = useRef(false);
  const hasServerOperationOccurred = useRef(false);

  useEffect(() => {
    if (!reduxIsLoading && accountWithdrawalSuccess && formSubmitted.current) {
      setIsLoading(false);
      closeModal();
      formSubmitted.current = false;
      return;
    }

    if (!reduxIsLoading && formSubmitted.current) {
      setIsLoading(false);
      formSubmitted.current = false;
    }
  }, [reduxIsLoading, accountWithdrawalSuccess, closeModal]);

  useEffect(() => {
    if (errorMessage && hasServerOperationOccurred.current) {
      setServerError(errorMessage);
    }
  }, [errorMessage]);

  const handleChange = (event) => {
    event.preventDefault();
    setValues({
      ...values,
      [event.target.name]: event.target.value,
    });
  };

  const handleClick = (event) => {
    event.preventDefault();
    const { isValid, errors } = validateCreateWithdrawal(values);
    setFormErrors({
      ...errors,
    });
    if (!isValid) {
      return;
    }

    setIsLoading(true);

    createWithdrawal({
      accountNo: values.accountNo,
      amount: values.amount,
      currency: values.currency,
    });

    formSubmitted.current = true;
    hasServerOperationOccurred.current = true;
  };

  const renderCreateForm = () => (
    <div className="withdraw-container">
        <div>
            <p>Make a withdrawal</p>
            <span className="error-text">{serverError}</span>
            <TextInput
                name="accountNo"
                placeholder="3245"
                title="Enter the account number to withdraw from"
                type="text"
                value={values.accountNo}
                error={formErrors.accountNo}
                onChange={handleChange} />
            <TextInput
                name="amount"
                placeholder="5000"
                title="Enter the amount to withdraw"
                type="text"
                value={values.amount}
                error={formErrors.amount}
                onChange={handleChange} />
            <Dropdown
                options={currencies}
                name="currency"
                optionsTitleProperty="name"
                optionsValueProperty="id"
                value={values.currency}
                error={formErrors.currency}
                onChange={handleChange}
                placeholder="select amount currency"
            />
            <br />
            <Button
                title="Withdraw"
                handleClick={handleClick}
                showLoader={isLoading}
                disabled={isLoading}
            />
        </div>
        <br />
    </div>
  );

  return (
        <div className="withdraw-account-container">
            <div className="form-container">
                    {renderCreateForm()}
                </div>
        </div>
  );
};

CreateWithdrawal.propTypes = {
  createWithdrawal: PropTypes.func.isRequired,
  closeModal: PropTypes.func.isRequired,
  isLoading: PropTypes.bool.isRequired,
  accountWithdrawalSuccess: PropTypes.bool.isRequired,
  errorMessage: PropTypes.string.isRequired,
};

const mapStateToProps = ({ account }) => ({
  isLoading: account.isWithdrawAccountLoading,
  accountWithdrawalSuccess: account.accountWithdrawalSuccess,
  errorMessage: account.errorMessage,
});

const mapDispatchToProps = {
  createWithdrawal: accountActions.createWithdrawal,
};

export default connect(mapStateToProps, mapDispatchToProps)(CreateWithdrawal);
