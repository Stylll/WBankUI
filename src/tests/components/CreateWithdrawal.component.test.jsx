import { render, fireEvent } from '@testing-library/react';
import React from 'react';

import { CreateWithdrawal } from '../../components/CreateWithdrawal/CreateWithdrawal.component';

describe('CreateWithdrawal Test', () => {
  it('should render properly', () => {
    const { baseElement } = render(<CreateWithdrawal
      createWithdrawal={() => {}}
      closeModal={() => {}}
      isLoading={false}
      accountWithdrawalSuccess={false}
      errorMessage=""
    />);
    expect(baseElement).toMatchSnapshot();
  });
  it('should validate the input', () => {
    const { baseElement, getByTestId, getByText } = render(<CreateWithdrawal
      createWithdrawal={() => {}}
      closeModal={() => {}}
      isLoading={false}
      accountWithdrawalSuccess={false}
      errorMessage=""
    />);
    fireEvent.click(getByTestId('createwithdrawal-withdraw'));
    expect(getByText('account number cannot be empty')).toBeInTheDocument();
    expect(getByText('amount cannot be empty')).toBeInTheDocument();
    expect(getByText('currency must be either cad, usd, or mxn')).toBeInTheDocument();
  });

  it('should submit form, call createWithdrawal and display error', () => {
    const createWithdrawal = jest.fn();
    const {
      baseElement, getByTestId, getByText, rerender,
    } = render(<CreateWithdrawal
      createWithdrawal={createWithdrawal}
      closeModal={() => {}}
      isLoading={false}
      accountWithdrawalSuccess={false}
      errorMessage=""
    />);

    fireEvent.change(getByTestId('createwithdrawal-accountNo'),
      { target: { value: '2345', name: 'accountNo' } });
    fireEvent.change(getByTestId('createwithdrawal-amount'),
      { target: { value: '245', name: 'amount' } });
    fireEvent.change(getByTestId('createwithdrawal-currency'),
      { target: { value: 'usd', name: 'currency' } });
    fireEvent.click(getByTestId('createwithdrawal-withdraw'));
    expect(createWithdrawal).toHaveBeenCalled();

    rerender(<CreateWithdrawal
      createWithdrawal={createWithdrawal}
      closeModal={() => {}}
      isLoading={false}
      accountWithdrawalSuccess={false}
      errorMessage="An error occurred"
    />);
    expect(getByText('An error occurred')).toBeInTheDocument();
  });
});
