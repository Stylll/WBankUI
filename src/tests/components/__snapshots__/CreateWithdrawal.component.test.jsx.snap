// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateWithdrawal Test should render properly 1`] = `
<body>
  <div>
    <div
      class="withdraw-account-container"
    >
      <div
        class="form-container"
      >
        <div
          class="withdraw-container"
        >
          <div>
            <p>
              Make a withdrawal
            </p>
            <span
              class="error-text"
            />
            <div
              class="form-input"
            >
              <div
                class="title"
              >
                Enter the account number to withdraw from
              </div>
              <input
                class="text-input"
                data-testid="createwithdrawal-accountNo"
                id=""
                name="accountNo"
                placeholder="3245"
                type="text"
                value=""
              />
              <span
                class="info"
              />
              
            </div>
            <div
              class="form-input"
            >
              <div
                class="title"
              >
                Enter the amount to withdraw
              </div>
              <input
                class="text-input"
                data-testid="createwithdrawal-amount"
                id=""
                name="amount"
                placeholder="5000"
                type="text"
                value=""
              />
              <span
                class="info"
              />
              
            </div>
            <div
              class="drop-down"
            >
              <select
                class="select"
                data-testid="createwithdrawal-currency"
                name="currency"
              >
                <option
                  value=""
                >
                  select amount currency
                </option>
                <option
                  value="cad"
                >
                  CAD
                </option>
                <option
                  value="usd"
                >
                  USD
                </option>
                <option
                  value="mxn"
                >
                  MXN
                </option>
              </select>
              <span
                class="chevy"
              >
                âˆ¨
              </span>
              <span
                class="info"
              />
              
            </div>
            <br />
            <div
              class="btn-button"
            >
              <button
                class="button "
                data-testid="createwithdrawal-withdraw"
                type="submit"
              >
                Withdraw
              </button>
            </div>
          </div>
          <br />
        </div>
      </div>
    </div>
  </div>
</body>
`;
